pkg/cvo/availableupdates.go

```diff
@@ -0,0 +1,156 @@
+package cvo
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/blang/semver"
+
+	"github.com/golang/glog"
+	"github.com/google/uuid"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openshift/cluster-version-operator/lib/resourcemerge"
+	cvv1 "github.com/openshift/cluster-version-operator/pkg/apis/config.openshift.io/v1"
+	osv1 "github.com/openshift/cluster-version-operator/pkg/apis/operatorstatus.openshift.io/v1"
+	"github.com/openshift/cluster-version-operator/pkg/cincinnati"
+)
+
+// syncAvailableUpdates attempts to retrieve the latest updates and update the status of the ClusterVersion
+// object. It will set the RetrievedUpdates condition. Updates are only checked if it has been more than
+// the minimumUpdateCheckInterval since the last check.
+func (optr *Operator) syncAvailableUpdates(config *cvv1.ClusterVersion) error {
+	upstream := optr.defaultUpstreamServer
+	if config.Spec.Upstream != nil {
+		upstream = string(*config.Spec.Upstream)
+	}
+	channel := config.Spec.Channel
+
+	// updates are only checked at most once per minimumUpdateCheckInterval or if the generation changes
+	u := optr.getAvailableUpdates()
+	if u != nil && u.Upstream == upstream && u.Channel == channel && u.RecentlyChanged(optr.minimumUpdateCheckInterval) {
+		glog.V(4).Infof("Available updates were recently retrieved, will try later.")
+		return nil
+	}
+
+	updates, condition := calculateAvailableUpdatesStatus(string(config.Spec.ClusterID), upstream, channel, optr.releaseVersion)
+	optr.setAvailableUpdates(&availableUpdates{
+		Upstream:  upstream,
+		Channel:   config.Spec.Channel,
+		Updates:   updates,
+		Condition: condition,
+	})
+	// requeue
+	optr.queue.Add(fmt.Sprintf("%s/%s", optr.namespace, optr.name))
```

Comment:
we are calling this `Sprintf` more than once, so maybe make it a func.

Reply:
yes, `opt.key()`